topology/probes/vpp/vpp.go

```diff
@@ -103,268 +124,511 @@ func (p *Probe) getInterfaceVrfID(ch api.Channel, index uint32) int64 {
-		needUpdate := make(map[uint32]struct{})
+	notifChan := make(chan api.Message, 10)
+	sub, err := ch.SubscribeNotification(notifChan, &interfaces.SwInterfaceEvent{})
+	if err != nil {
+		p.Ctx.Logger.Error(err)
+		return
+	}
 
-		req := &interfaces.SwInterfaceDump{}
-		reqCtx := ch.SendMultiRequest(req)
-		for {
-			msg := &interfaces.SwInterfaceDetails{}
-			stop, err := reqCtx.ReceiveReply(msg)
-			if stop {
-				break
-			}
-			if err != nil {
-				p.Ctx.Logger.Error(err)
-				goto nextEvents
-			}
+	p.Ctx.Logger.Debugf("Registering for VPP events")
+	p.interfaceEventsEnableDisable(ch, true)
 
-			intf, found := p.interfaceMap[msg.SwIfIndex]
-			if !found || interfaceNeedUpdate(msg, intf) {
-				needUpdate[msg.SwIfIndex] = struct{}{}
-			}
-			if found {
-				foundInterfaces[msg.SwIfIndex] = struct{}{}
+	if err := p.synchronize(ch); err != nil {
+		p.Ctx.Logger.Error(err)
+		return
+	}
+
+	pollingTicker := time.NewTicker(VPPPollingTime)
+	defer pollingTicker.Stop()
+
+	pingTimer := time.NewTimer(HealthCheckInterval)
+	defer pingTimer.Stop()
+
+LOOP:
+	for {
+		select {
+		case notif := <-notifChan:
+			msg := notif.(*interfaces.SwInterfaceEvent)
+			p.Ctx.Logger.Debugf("Received sw interface event %+v", msg)
+			p.handleInterfaceEvent(msg)
+		case <-pollingTicker.C:
+			err = p.synchronize(ch)
+		case <-pingTimer.C:
+			req := &vpe.ControlPing{}
+			msg := &vpe.ControlPingReply{}
+			if err = ch.SendRequest(req).ReceiveReply(msg); err != nil {
+				err = fmt.Errorf("Health check failed: %s", err)
 			}
-			p.interfaceMap[msg.SwIfIndex] = msg
+		case <-ctx.Done():
+			defer p.conn.Disconnect()
+			break LOOP
 		}
 
-		/* Update interface metadata */
-		for index := range needUpdate {
-			msg := p.interfaceMap[index]
-			p.Ctx.Logger.Debugf("Add/Update interface %v idx %d up/down %d", strings.Trim(string(msg.InterfaceName), "\000"), int64(msg.SwIfIndex), int64(msg.AdminUpDown))
-			p.eventAddInterface(ch, msg)
-		}
-		/* Remove interface that didn't exist anymore */
-		for index := range p.interfaceMap {
-			_, found := foundInterfaces[index]
-			_, firsttime := needUpdate[index]
-			if !found && !firsttime {
-				node := p.getInterface(index)
-				p.Ctx.Graph.RLock()
-				name, _ := node.GetFieldString("Name")
-				p.Ctx.Graph.RUnlock()
-				p.Ctx.Logger.Debugf("Delete interface %v idx %d", name, index)
-				p.eventDelInterface(node)
-				delete(p.interfaceMap, index)
-			}
+		if err != nil {
+			p.Ctx.Logger.Error(err)
+			break
 		}
 
-	nextEvents:
-		time.Sleep(VPPPollingTime * time.Millisecond)
+		pingTimer.Reset(HealthCheckInterval)
 	}
 
+	p.Ctx.Logger.Debugf("Unregistering for VPP events")
+	p.interfaceEventsEnableDisable(ch, false)
+
+	sub.Unsubscribe()
 	ch.Close()
 }
 
-// Start VPP probe and get all interfaces
-func (p *Probe) Start() error {
-	conn, err := govpp.Connect(p.shm)
+// Do starts the VPP probe and get all interfaces
+func (p *Probe) Do(ctx context.Context, wg *sync.WaitGroup) error {
+	conn, err := govpp.Connect(p.addr)
```

Comment:
I'm getting lot's of these error messages for the host and containers that do not run any VPP:
```
skydive-agent_1     | 2019-10-17T12:23:27.844Z	ERROR	probes/wrapper.go:52 (*serviceManager).Start.func1	dev-VirtualBox: VPP connection error: VPP API socket file /run/vpp/api.sock does not exist
skydive-agent_1     | 2019-10-17T12:23:28.016Z	ERROR	probes/wrapper.go:52 (*serviceManager).Start.func1	vpp-9d93ad24-9cc6-4d85-584e-c333eb1777ae: VPP connection error: VPP API socket file /run/vpp/api.sock does not exist
```

Perhaps `Do()` method could start by checking if the socket files exists or if a process named `vpp` exists and continue only if it is the case to avoid polluting the log.